Assets/MixedRealityToolkit.SDK/Experimental/Features/Support/Scripts/BaseServiceManager.cs

```diff
@@ -0,0 +1,395 @@
+        /// </summary>
+        protected Dictionary<Type, IMixedRealityService> registeredServices = new Dictionary<Type, IMixedRealityService>();
+
+        /// <summary>
+        /// The collection of registered data providers.
+        /// </summary>
+        private List<IMixedRealityDataProvider> dataProviders = new List<IMixedRealityDataProvider>();
+
+        #region MonoBehaviour implementation
+
+        protected virtual void Update()
+        {
+            if (Application.isPlaying)
+            {
+                IReadOnlyList<IMixedRealityService> serviceSnapshot = new List<IMixedRealityService>(registeredServices.Values);
+                for (int i = 0; i < serviceSnapshot.Count; i++)
+                {
+                    serviceSnapshot[i]?.Update();
+                }
+
+                IMixedRealityDataProvider[] providers = dataProviders.ToArray();
+                for (int i = 0; i < providers.Length; i++)
+                {
+                    providers[i]?.Update();
+                }
+            }
+        }
+        private void LateUpdate()
+        {
+            if (Application.isPlaying)
+            {
+                IReadOnlyList<IMixedRealityService> serviceSnapshot = new List<IMixedRealityService>(registeredServices.Values);
+                for (int i = 0; i < serviceSnapshot.Count; i++)
+                {
+                    serviceSnapshot[i]?.LateUpdate();
+                }
+
+                IMixedRealityDataProvider[] providers = dataProviders.ToArray();
+                for (int i = 0; i < providers.Length; i++)
+                {
+                    providers[i]?.LateUpdate();
+                }
+            }
+        }
+
+
+        protected virtual void OnEnable()
+        {
+            if (Application.isPlaying)
+            {
+                IReadOnlyList<IMixedRealityService> serviceSnapshot = new List<IMixedRealityService>(registeredServices.Values);
+                for (int i = 0; i < serviceSnapshot.Count; i++)
+                {
+                    serviceSnapshot[i]?.Enable();
+                }
+
+                IMixedRealityDataProvider[] providers = dataProviders.ToArray();
+                for (int i = 0; i < providers.Length; i++)
+                {
+                    providers[i]?.Enable();
+                }
+            }
+        }
+
+        protected virtual void OnDisable()
+        {
+            if (Application.isPlaying)
+            {
+                IMixedRealityDataProvider[] providers = dataProviders.ToArray();
+                for (int i = 0; i < providers.Length; i++)
+                {
+                    providers[i]?.Disable();
+                }
+
+                IReadOnlyList<IMixedRealityService> serviceSnapshot = new List<IMixedRealityService>(registeredServices.Values);
+                for (int i = 0; i < serviceSnapshot.Count; i++)
+                {
+                    serviceSnapshot[i]?.Disable();
+                }
+            }
+        }
```

Comment:
Same general comments about gc alloc and null checking

Reply:
fixed